src/JsonPath/FunctionRepository.cs

```diff
@@ -16,44 +16,107 @@ public static class FunctionRepository
 
 	static FunctionRepository()
 	{
-		Register(new LengthFunction());
-		Register(new CountFunction());
-		Register(new MatchFunction());
-		Register(new SearchFunction());
-		Register(new ValueFunction());
+		RegisterValueFunction<LengthFunction>();
+		RegisterValueFunction<CountFunction>();
+		RegisterLogicalFunction<MatchFunction>();
+		RegisterLogicalFunction<SearchFunction>();
+		RegisterValueFunction<ValueFunction>();
 	}
 
 	/// <summary>
 	/// Registers a new function implementation, allowing it to be parsed.
 	/// </summary>
 	/// <param name="function">The function.</param>
+	[RequiresUnreferencedCode("Use the overloads that take a Type for AOT compatibility")]
 	public static void Register(ValueFunctionDefinition function)
+	{
+		Register(function, function.GetType());
+	}
+
+	/// <summary>
+	/// Registers a new function implementation, allowing it to be parsed.
+	/// </summary>
+	/// <param name="function">The function.</param>
+	/// <param name="functionType">The type of the function.</param>
+	public static void Register(ValueFunctionDefinition function,
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type functionType)
```

Comment:
I don't understand how this fixes the problem.  Isn't this marking the method (via the param) as non-AOT?

Reply:
Reflection is compatible with AOT (all metadata for all types is included by default), but not compatible with trimming. Trimming is another layer on top of AOT where you can reduce the size of the final binary by trimming code/data out that the compiler can tell is unreachable. For reflection, metadata is trimmed out by default _unless_ the compiler can see statically that that some code needs the metadata.

These attributes (as described [here](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming)) are how you annotate `Type` parameters or template arguments with which aspects of metadata the function is going to need.